docs/src/references/agent/device-profile.md

```diff
@@ -0,0 +1,327 @@
+                "version": "latest",
+                "action": "install"
+              },
+              {
+                "name": "collectd",
+                "version": "latest",
+                "type": "apt",
+                "action": "install"
+              }
+            ]
+          }
+        ]
+      }
+    },
+    {
+      "operation": "config_update",
+      "skip": false,
+      "payload": {
+        "type": "collectd.conf",
+        "remoteUrl":"https://abc.com/some/collectd/conf",
+        "path": "/etc/collectd/collectd.conf"
+      }
+    },
+    {
+      "operation": "software_update",
+      "skip": false,
+      "payload": {
+        "updateList": [
+          {
+            "type": "apt",
+            "modules": [
+              {
+                "name": "jq",
+                "version": "latest",
+                "action": "install"
+              }
+            ]
+          }
+        ]
+      }
+    }
+  ]
+}'
+```
+
+The profile definition in the payload is an array of operations.
+Each operation could be a `firmware_update`, `software_update` or `config_update`.
+The operations are executed in the order in which they are defined in the profile definition, by default.
+There is no restriction on the order of modules in a profile and hence can be defined in any preferred order.
+For example, additional software can be installed or configurations updated before the firmware is updated.
+This default execution order can also be overridden in the workflow definition, by updating the order in the `scheduled` state.
+
+The `"skip"` field is optional and the value is `false`, by default.
+It can be used to skip any operations during the development/testing phase, without fully deleting the entry from the profile.
+
+## Tedge agent handling device profile commands
+
+The `tedge-agent` handles `device_profile` commands using the workflow definition at `/etc/tedge/operations/device_profile.toml`.
+This workflow definition handles each module type using sub-operation workflows defined for that type.
+For example, the firmware module is installed by triggering a `firmware_update` sub-command
+which in turn uses the `firmware_update` workflow for that operation execution.
+Similarly software modules are installed with `software_update` subcommands and
+configuration updates are applied using `config_update` subcommands.
+These subcommands are triggered for each module defined in the profile definition in that order.
+
+Here is a sample device profile workflow:
+
+```toml
+operation = "device_profile"
+
+[init]
+action = "proceed"
+on_success = "scheduled"
+
+# Sort the inputs as desired
+[scheduled]
+script = "/etc/tedge/operations/device_profile.sh ${.payload.status} ${.payload}"
+on_success = "executing"
+on_error = { status = "failed", reason = "fail to sort the profile list"}
+
+[executing]
+action = "builtin"
+on_success = "next_operation"
+
+[next_operation]
+action = "iterator"
```

Comment:
I went with this new `iterator` action instead of `builtin` just to provide a clear `on_next` handle.

Reply:
> I went with this new `iterator` action instead of `builtin` just to provide a clear `on_next` handle.

This is clearer for the user and also more consistent with the current implementation. Using a `builtin` action here would be confusing as it would mean "delegates the step (here `next_operation`) to a built-in implementation of the workflow (here `device-profile`)".